core/src/main/java/io/temporal/samples/earlyreturn/TransactionWorkflowImpl.java

```diff
@@ -0,0 +1,70 @@
+/*
+ *  Copyright (c) 2020 Temporal Technologies, Inc. All Rights Reserved
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.samples.earlyreturn;
+
+import io.temporal.activity.ActivityOptions;
+import io.temporal.workflow.Workflow;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TransactionWorkflowImpl implements TransactionWorkflow {
+  private static final Logger log = LoggerFactory.getLogger(TransactionWorkflowImpl.class);
+  private final TransactionActivities activities =
+      Workflow.newActivityStub(
+          TransactionActivities.class,
+          ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofSeconds(30)).build());
+
+  private boolean initDone = false;
+  private Transaction tx;
+  private Exception initError = null;
+
+  @Override
+  public TxResult processTransaction(Transaction txInput) {
+    this.tx = txInput;
```

Comment:
I would refactor this sample to make `Transaction` immutable and not modify transaction as the workflow progress. I would make two different classes one for the workflow input that doesn't have a transaction ID field and one  (maybe returned by the activity `initTransaction` or created in the workflow) that the `cancelTransaction` and `completeTransaction`

Reply:
Done: creation of a TransactionRequest object + the Transaction class is now immutable